core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/uni/IfExistsUniWithUniNode.java

```diff
@@ -78,13 +77,10 @@ public void insertA(UniTuple<A> tupleA) {
         Counter<A> counter = new Counter<>(tupleA);
         indexerA.put(indexProperties, tupleA, counter);
 
-        Map<UniTuple<B>, Set<Counter<A>>> counterSetMapB = indexerB.get(indexProperties);
-        counter.countB = counterSetMapB.size();
+        counter.countB = indexerB.countValues(indexProperties);
         if (shouldExist ? counter.countB > 0 : counter.countB == 0) {
             counter.state = BavetTupleState.CREATING;
-            for (Set<Counter<A>> counterSetB : counterSetMapB.values()) {
-                counterSetB.add(counter);
-            }
+            indexerB.visit(indexProperties, (tuple, counterSetB) -> counterSetB.add(counter));
```

Comment:
This is not nice, as it forces us to first count (1 iteration) then visit (second iteration).
Therefore this "optimization" unfortunately results in the exact same behavior as before, only this time with much less memory used in the process.

Reply:
My local changes change this code anyway. We might need an extra method on the Indexer to allow this to be done efficiently.

Counting (without filtering) for IfExistsNode should probably not rely on the same visit code of JoinNode, because counting can be done more efficiently.